### PR TITLE
feat: add source_invite_code to guildmemberdata

### PR DESCRIPTION
**Pull request by LawMixer**
Drafting this pull request until it is officially added to Discord API, though it is soft released just not announced. Will undraft it when it is timed to merge it.

Link: https://github.com/discord/discord-api-docs/discussions/3210#discussioncomment-7503972

### CODE REVIEW

**Review by silasary — APPROVED** [review]

**Comment by LawMixer** [comment]
I'm ready whenever the maintainers are, everything should be completed! Thank you!

File: discord_typings/_resources/_guild.py
Comment (Bluenix2) [inline discord_typings/_resources/_guild.py]: As shown in the discussion you linked, this can be null, which means this should be marked as possibly None.

Additionally, due to how it is seemingly only passed in one endpoint, and because it is unofficial, I would recommend marking it as possibly missing. This also means it will likely interfere less with existing code.

```suggestion
    source_invite_code: NotRequired[Optional[str]]
```

**Review by Bluenix2 — CHANGES_REQUESTED** [review]
Hi, I've been thinking about this pull request a bit, and waiting to hear about any official news.

Thanks a lot for your first PR! :heart: Do you know more about the nature of this new field than the discussion you linked?

I am hesitant to add unofficial fields because they can quickly grow out of hand, but that isn't to say that this library follows the docs religiously. It deviates from the docs when doing so improves the typing experience (such as in the case of channels). However, I don't believe this field matches that criteria.

Because it isn't in the official docs, and a [quick search on `discord-api-types` didn't give any results](https://github.com/search?q=repo%3Adiscordjs%2Fdiscord-api-types%20source_invite_code&type=code) (as of writing), I would like to hold off on merging this unless there is a big need to.

Thanks again!

**Comment by LawMixer** [comment]
This change has been slowly coming out for all users, though it seems Discord has not made any pull request for their API documentations for this. Thank you for having me! <3 

I just committed a new version with your changes, let me know if there's anything else I can do!

**Comment by LawMixer** [comment]
> Thanks a lot for your first PR! ❤️ Do you know more about the nature of this new field than the discussion you linked?

No yet announced from Discord officially, but we've seen ``source_invite_code` `show up in the Members page on people's discord server, and it is in the member search endpoint. 

We've gotten an update from [Discord Support on X](https://twitter.com/discord_support/status/1730267160906264851) that pushed things along wayyy quicker, so if you want to wait on them for the official announcement of this change, that's fine by me.

File: discord_typings/_resources/_guild.py
Comment (MCausc78) [inline discord_typings/_resources/_guild.py]: This isnt extra key for GuildMember
```json
{
    "member": {
        "avatar": null,
        "banner": null,
        "communication_disabled_until": null,
        "flags": 0,
        "joined_at": "2024-06-22T22:42:07.773000+00:00",
        "nick": null,
        "pending": false,
        "premium_since": null,
        "roles": [
            "1233367105710063616",
            "1254204805891625134"
        ],
        "unusual_dm_activity_until": null,
        "user": {
            "id": "270904126974590976",
            "username": "Dank Memer",
            "avatar": "a_24778db4737114253ac3b30f45f1979f",
            "discriminator": "5192",
            "public_flags": 589824,
            "flags": 589824,
            "bot": true,
            "banner": null,
            "accent_color": null,
            "global_name": null,
            "avatar_decoration_data": null,
            "banner_color": null,
            "clan": null
        },
        "mute": false,
        "deaf": false
    },
    "source_invite_code": null,
    "join_source_type": 0,
    "inviter_id": null
}
```
For `join_source_type`, there is a [enum](https://github.com/discord/discord-api-docs/pull/7068/files#diff-6a43fc10e10de61f31b90ab22421ef0612d0ea75e3cd53891c6393a6fb84872dR408-R419).

Comment (LawMixer) [reply discord_typings/_resources/_guild.py]: I've just resolved this issue, could you review it again please and thank you?

File: discord_typings/_resources/_guild.py
Comment (MCausc78) [inline discord_typings/_resources/_guild.py]: `BOT = 1,` makes enum value to be `tuple[int]`, remove comma

File: discord_typings/_resources/_guild.py
Comment (MCausc78) [inline discord_typings/_resources/_guild.py]: These fields should be in new `SupplementalGuildMemberData` type instead

`source_invite_code`, and `inviter_id` must be Optional since member might be a bot

Comment (LawMixer) [reply discord_typings/_resources/_guild.py]: Good catch!

File: discord_typings/_resources/_guild.py
Comment (LawMixer) [reply discord_typings/_resources/_guild.py]: Good catch!

**Comment by Bluenix2** [comment]
Hi again!

I appreciate the effort put into this, but considering it has been quite some time since the PR was closed by Discord I don't think the future looks very bright for this feature.

Some piece of feedback on the code here:

* Because of how Python does enums with special Enum metaclasses, you will also notice all other "Types" are in reality Literals
* The following piece of code really does not make sense, and will fail:

```python
    join_source_type: JoinSourceType[int]
```

Thanks @MCausc78 for having a look at this as well ❤️

/ Blue

**Pull request closed without merging**
